### PR TITLE
implement the retry integrity tag

### PR DESCRIPTION
**Pull request by marten-seemann**
See https://github.com/quicwg/base-drafts/pull/3120.

### CODE REVIEW

File: internal/handshake/retry.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"fmt"
+
+	"github.com/lucas-clemente/quic-go/internal/protocol"
+)
+
+var retryAEAD cipher.AEAD
```
Comment (lucas-clemente) [inline internal/handshake/retry.go]: cipher.AEAD doesn't seem to make any thread safety guarantees. I think you might need a mutex here, or make this a server local var?

Comment (marten-seemann) [reply internal/handshake/retry.go]: I'm not sure I understand what you mean. `retryAEAD` is only set in `init()`, after we just call `Seal`.
`Seal` doesn't modify any internal state (at least not that I'm aware of), so it should be safe to use it concurrently, right?

Comment (lucas-clemente) [reply internal/handshake/retry.go]: As long as that’s not guaranteed by the interface or the constructor, you can’t make that assumption.

**Comment by marten-seemann** [comment]
@lucas-clemente I added a mutex around the retry AEAD. This allows us to use pre-allocate some global state, which leads to considerably fewer allocations as well as a significant speedup (in the non-contended case):
```
pkg: github.com/lucas-clemente/quic-go/internal/handshake
BenchmarkRetry-16       	15345738	        71.8 ns/op	      16 B/op	       1 allocs/op
BenchmarkRetryOld-16    	 9577706	       112 ns/op	      56 B/op	       4 allocs/op

**Pull request merged**